src/Microsoft.IdentityModel.Tokens/Encryption/RsaKeyWrapProvider.cs

```diff
@@ -0,0 +1,459 @@
+        /// <exception cref="ArgumentNullException">'key' is null.</exception>
+        /// <exception cref="ArgumentNullException">'algorithm' is null.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">The keysize doesn't match the algorithm.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">The <see cref="SecurityKey"/>  is not supported.</exception>
+        /// <exception cref="ArgumentException">If <see cref="SecurityKey"/> and algorithm pair are not supported.</exception>
+        /// <exception cref="InvalidOperationException">Failed to create RSA algorithm with provided key and algorithm.</exception>
+        /// </summary>
+        public RsaKeyWrapProvider(SecurityKey key, string algorithm, bool willDecrypt)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            if (string.IsNullOrEmpty(algorithm))
+                throw LogHelper.LogArgumentNullException(nameof(algorithm));
+
+            if (key.KeySize < 2048)
+            {
+                string keyId = key.KeyId ?? string.Empty;
+                throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException("key.KeySize", string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10662, algorithm, 2048, keyId, key.KeySize)));
+            }
+
+            if (!IsSupportedAlgorithm(key, algorithm))
+                throw LogHelper.LogExceptionMessage(new ArgumentException(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10671, algorithm, key)));
+
+            Algorithm = algorithm;
+            Key = key;
+
+            ResolveRsaAlgorithm(key, algorithm, willDecrypt);
+#if NETSTANDARD1_4
+            if (_rsa == null && _rsaCryptoServiceProviderProxy == null)
+                throw LogHelper.LogExceptionMessage(new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10672)));
+#else
+            if (_rsaCryptoServiceProvider == null && _rsaCryptoServiceProviderProxy == null)
+                throw LogHelper.LogExceptionMessage(new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10672)));
+#endif
+        }
+
+        /// <summary>
+        /// Gets the KeyWrap algorithm that is being used.
+        /// </summary>
+        public string Algorithm { get; private set; }
+
+        /// <summary>
+        /// Gets or sets a user context for a <see cref="KeyWrapProvider"/>.
+        /// </summary>
+        /// <remarks>This is null by default. This can be used by runtimes or for extensibility scenarios.</remarks>
+        public string Context { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="SecurityKey"/> that is being used.
+        /// </summary>
+        public SecurityKey Key { get; private set; }
+
+        private RSAParameters CreateRsaParametersFromJsonWebKey(JsonWebKey webKey, bool isPrivate)
```

Comment:
We should consolidate this into a function that can be accessed from different locations. A similar function seems to be found here: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/Microsoft.IdentityModel.Tokens/AsymmetricSignatureProvider.cs#L383